### ISSUE TITLE
Compare Symbol memory addresses

### ISSUE THREAD

**Issue by certik** [sympy/csympy#379]
Instead of comparing the `std::string`.

Pros: That speeds up things.

Cons: The disadvantage is that now two `Symbol("x")` created separately will compare as false.

Preliminary implementation in #124.

**Comment by shivamvats** [comment]
We mostly create `symbols` with unique names.  So, the con shouldn't really be a major issue.

+1

**Comment by shivamvats** [comment]
Btw, let's say I do the following:
```
RCP<Basic> x = symbol('x');
RCP<Basic> y = symbol('x');
```
Now instead of creating a new object with a separate memory location for y, we could just create a reference to x in y. That should solve the problem of comparing two `symbols` with the same string.

**Comment by certik** [comment]
On Fri, Jan 30, 2015 at 10:36 PM, Shivam Vats <notifications@github.com>
wrote:

> Btw, let's say I do the following:
>
> RCP<Basic> x = symbol('x');
> RCP<Basic> y = symbol('x');
>
> Now instead of creating a new object with a separate memory location for
> y, we could just create a reference to x in y. That should solve the
> problem of comparing two symbols with the same string.
>

But how would you do that? Some global hashtable of symbols?

Ondrej

**Comment by shivamvats** [comment]
Yes. We could have a table of all the `symbols` that have the same name/string. Once we have that, I think comparing just the hash should also work.

**Comment by certik** [comment]
The only disadvantage of a global table of symbols is that it can grow very big, so we need to manage that somehow. Perhaps when the symbol goes out of scope, we need to somehow figure out how to deallocate it and remove it from the hashtable.

**Comment by shivamvats** [comment]
We could try something similar to the reference counting mechanism by adding a `count` to every string.

**Comment by certik** [comment]
Or we can just check the refcount in RCP, if it is 1, then it ran out of scope and the only reference left is in the hashtable, so it should get deallocated.

**Comment by shivamvats** [comment]
Right! Let me try a rough implementation and see how it goes.

**Comment by shivamvats** [comment]
Btw, we can also use the hash values to compare symbols. I am not sure how much of a speed advantage that gives. I tried running the benchmark `expand2` but the time taken varies every time I run the same benchmark with the same code.

**Comment by shivamvats** [comment]
@certik I benchmarked hash comparisons(https://github.com/shivamvats/csympy/commit/ab272fd37d5b7642ff914c05daea490a06214dbf), string comparisons(master) and memory address comparison(your pr). Hash comparisons seem to be the fastest:
```
Hash_comparison:
4249ms
4223ms
4205ms
4235ms
4252ms

String comparison:
4465ms
4549ms
4552ms
4663ms
4494ms

Memory comparison:
4570ms
4548ms
4528ms
4532ms
4569ms
```
